ibis/backends/dask/execution/util.py

```diff
@@ -345,3 +348,106 @@ def safe_scalar_type(output_meta):
         output_meta = pd.Timestamp(1, tz=output_meta.tz, unit=output_meta.unit)
 
     return output_meta
+
+
+def add_partitioned_sorted_column(
+    df: Union[dd.DataFrame, dd.Series],
+    set_index: bool = False,
+    col_name: str = "_ibis_index",
+) -> Tuple[dd.DataFrame, List]:
+    """Add a column that is already partitioned and sorted
+    This columns acts as if we had a global index across the distributed data.
+    Important properties:
+    - Each row has a unique id (i.e. a value in this column)
+    - IDs within each partition are already sorted
+    - Any id in partition N_{t} is less than any id in partition N_{t+1}
+    We do this by designating a sufficiently large space of integers per
+    partition via a base and adding the existing index to that base. See
+    `helper` below.
+    Though the space per partition is bounded, real world usage should not
+    hit these bounds. We also do not explicity deal with overflow in the
+    bounds.
+    Parameters
+    ----------
+    df : dd.DataFrame
+        Dataframe to add a column to
+    set_index : bool, optional
+        Whether to also set that column as the index, by default False
+    col_name : str, optional
```

Comment:
don't you always make a unique name? (e.g. uuid0?

Reply:
I don't think we want a unique name - it is useful to be able to refer to the column downstream (this will be more important in windows grouped analytic udf code where I'll need to re-index on `_ibis_index` after performing an operation on groups. 

I can raise a `ValueError` if a user calls this more than once (i.e. raise if `col_name in df.columns`).

Reply:
My understanding is that the index column is "temporarily". If so, what is the scope of the index column? E.g., when will this be dropped?

Reply:
We always drop the index as part of `execute_and_reset` (same as the pandas backend) (https://github.com/ibis-project/ibis/blob/b9556084c0b99b2b5dedcb6efb94472ff4da19c8/ibis/backends/dask/core.py#L455) I can drop this earlier as part of the select statement, but that feels unnecessary.

Reply:
I think there is a major difference here, when dropped inside `execute_and_reset`, it is done only after the whole tree is executed, so the scope of the index is larger and we need to think more about implications of this column through out the entire tree. Whereas if we drop the column after selection, then the scope is much smaller and easier to reason about.

Reply:
To elaborate a bit more, the reason that I think this is better is because passing around an "motorically increasing index" but not maintaining it in every op is dangerous. E.g., if user did some operations that breaks the index (e.g., sort), there is currently no way that we can guarantee the " motorically increasing index invariance" across the full life time of an execution.

Reply:
I've now changed `add_partitioned_sorted_column`:

1. Always `set_index`
2. Removed the ability to pass in the name of the column
3. Do not pass back divisions
4. Drop this index after selection is done. 

This simplifies the current state of this. I'll re-add if I need these features later.